flink-optimizer/src/main/java/org/apache/flink/optimizer/traversals/RangePartitionRewriter.java

```diff
@@ -0,0 +1,188 @@
+		sipPlanNode.setParallelism(sourceParallelism);
+		sipChannel.setTarget(sipPlanNode);
+		appendOutgoingChannels.add(sipChannel);
+		this.plan.getAllNodes().add(sipPlanNode);
+
+		// 2. Fixed size sample in a single coordinator.
+		SampleInCoordinator sampleInCoordinator = new SampleInCoordinator(false, sampleSize, seed);
+		UnaryOperatorInformation sicOperatorInformation = new UnaryOperatorInformation(isdTypeInformation, sourceOutputType);
+		GroupReduceOperatorBase sicOperatorBase = new GroupReduceOperatorBase(sampleInCoordinator, sicOperatorInformation, "Sample in coordinator");
+		GroupReduceNode sicNode = new GroupReduceNode(sicOperatorBase);
+		Channel sicChannel = new Channel(sipPlanNode, TempMode.NONE);
+		sicChannel.setShipStrategy(ShipStrategyType.PARTITION_HASH, channel.getShipStrategyKeys(), channel.getShipStrategySortOrder(), null, channel.getDataExchangeMode());
+		SingleInputPlanNode sicPlanNode = new SingleInputPlanNode(sicNode, "SampleInCoordinator PlanNode", sicChannel, DriverStrategy.ALL_GROUP_REDUCE);
+		sicPlanNode.setParallelism(1);
+		sicChannel.setTarget(sicPlanNode);
+		sipPlanNode.addOutgoingChannel(sicChannel);
+		this.plan.getAllNodes().add(sicPlanNode);
+
+		// 3. Use sampled data to build range boundaries.
+		RangeBoundaryBuilder rangeBoundaryBuilder = new RangeBoundaryBuilder(comparator, targetParallelism);
+		TypeInformation<CommonRangeBoundaries> rbTypeInformation = TypeExtractor.getForClass(CommonRangeBoundaries.class);
+		UnaryOperatorInformation rbOperatorInformation = new UnaryOperatorInformation(sourceOutputType, rbTypeInformation);
+		MapPartitionOperatorBase rbOperatorBase = new MapPartitionOperatorBase(rangeBoundaryBuilder, rbOperatorInformation, "RangeBoundaryBuilder");
+		MapPartitionNode rbNode= new MapPartitionNode(rbOperatorBase);
+		Channel rbChannel = new Channel(sicPlanNode, TempMode.NONE);
+		rbChannel.setShipStrategy(ShipStrategyType.FORWARD, channel.getDataExchangeMode());
+		SingleInputPlanNode rbPlanNode = new SingleInputPlanNode(rbNode, "RangeBoundary PlanNode", rbChannel, DriverStrategy.MAP_PARTITION);
+		rbPlanNode.setParallelism(1);
+		rbChannel.setTarget(rbPlanNode);
+		sicPlanNode.addOutgoingChannel(rbChannel);
+		this.plan.getAllNodes().add(rbPlanNode);
+
+		// 4. Take range boundaries as broadcast input and take the tuple of partition id and record as output.
+		AssignRangeIndex assignRangeIndex = new AssignRangeIndex(comparator);
+		TypeInformation<Tuple2> ariOutputTypeInformation = new TupleTypeInfo<>(BasicTypeInfo.INT_TYPE_INFO, sourceOutputType);
+		UnaryOperatorInformation ariOperatorInformation = new UnaryOperatorInformation(sourceOutputType, ariOutputTypeInformation);
+		MapPartitionOperatorBase ariOperatorBase = new MapPartitionOperatorBase(assignRangeIndex, ariOperatorInformation, "Assign Range Index");
+		MapPartitionNode ariNode= new MapPartitionNode(ariOperatorBase);
+		Channel ariChannel = new Channel(sourceNode, TempMode.NONE);
```

Comment:
OK, so `DataExchangeMode` is a new concept and should replace `TempMode`. However, the transition is not complete yet. The difference is that `TempMode` temps data within the operator and `DataExchangeMode` temps in the network stack. So, the correct channel configuration is to set the `TempMode` for both source-node outgoing channels to `TempMode.NONE` and the `DataExchangeMode` for the sampling out-channel to `DataExchangeMode.PIPELINED` and for the assignment out-channel to `DataExchangeMode.BATCH`.